### PR TITLE
Generalized image upload

### PR DESCRIPTION
**Pull request by WilhelmOks**
To support the upload of gifs, or any other kind of animated images, I've changed the type of the image parameters from UIImage to Data.

I removed all the duplicated versions of functions for mac which took NSImage as a parameter, because they are also of type Data now and are (or should be) equal to the iOS functions.

This is obviously a breaking change, but on my SwiftUI client, the change was really minimal because I have the images in Data objects anyway. But even if you need to encode UIImage/NSImage to Data, you can just use the same functions that SwiftRant has used.
I would argue that it is the client's resposibility anyway. For instance, it should decide what jpeg quality it wants the images to be encoded in before uploading. And of course, it gives clients the flexibility to upload different image formats.

### CODE REVIEW

**Comment by OmerFlame** [comment]
This is a massive change, I will have to take a closer, more careful look at this but I can't this Friday and Saturday. I will look on Sunday though!

**Comment by WilhelmOks** [comment]
Ok, thanks for letting me know.

**Comment by OmerFlame** [comment]
This pull request is not exactly super urgent, as this doesn't fix anything. I do see the reason for it, but it is not absolutely essentially right now. I will keep this hanging a little for now until I have time to review this and also make the necessary changes to my AltRant project before going with this. This will stay open in the case that there is a need to fix things while it's waiting for approval.

**Comment by WilhelmOks** [comment]
Ok, I just want to add that even though it looks massive, the actual change is really not big:

* changing UIImage to Data
* skipping conversion from UIImage to Data, because it is Data already.
* deleting redundant code

I switched to this branch for JoyRant temporarily until it is merged, because I want to support GIF upload.

So, yes. It's not urgent but still important, imo.

**Comment by WilhelmOks** [comment]
I made an addition to this PR:

Reason:
Before this change, we can pick images for upload that are not supported by devRant.
For example, when we pick a webp image and upload it using the devRant api, it will return an error.
I want that it is also possible to upload such images.

Solution:
Before uploading the image data, SwiftRant will automatically check the image Data format if it is supported and will make a conversion to JPEG if necessary.

Details of implementation:
I've implemented an extensible system of image data converters.
Each image uploading function takes an array of converters.
The default value is only one converter which converts unsupported formats to JPEG.
So the function on the call side doesn't change. It will do a reasonable conversion by default if needed.
However, an empty array can be passed for the converters to disable conversion (if someone needs it for whatever reason).
And custom converters can be written by users of SwiftRant and passed into the converter list. 
The converters will be applied in sequence.

**Comment by WilhelmOks** [comment]
Friendly reminder

**Comment by OmerFlame** [comment]
I know, sorry for not working on this. I am really busy lately and I can't work on this at the moment. I will have more free time in the coming week or two probably.

**Comment by OmerFlame** [comment]
I did not forget about this, I finally have a little bit of time to be a bit more active here.

Can you give an example for how you would use the conversion system/just send a whole post with the proposed changes?

**Comment by WilhelmOks** [comment]
Using the defaults, the new call wouldn't look any different:
`try await swiftRant.postRant(try token(), postType: type, content: content, tags: tags, image: image).get()`
with image being Data instead of UIImage/NSImage.
The default value for the `imageConversion` is `[.unsupportedToJpeg]` implicitly so this is equivalent to
`try await swiftRant.postRant(try token(), postType: type, content: content, tags: tags, image: image, imageConversion: [.unsupportedToJpeg]).get()`

Let's suppose someone wants to use the conversion system to convert jpeg to png, in addition to the default conversion.
The converter could be implemented like this:
```
struct JpegToPngConverter: ImageDataConverter {
    func convert(_ data: Data) -> Data {
        switch data.first {
        case 0xFF: // this is jpeg. convert to png.
            return UIImage(data: data)?.pngData() ?? data
        default: // don't convert any other formats
            return data
        }
    }
}

// this is just for convenience
extension ImageDataConverter where Self == JpegToPngConverter {
    static var jpegToPng: Self { Self() }
}
```

Then the post rant call would look like this:
`try await swiftRant.postRant(try token(), postType: type, content: content, tags: tags, image: image, imageConversion: [.unsupportedToJpeg, .jpegToPng]).get()`

First, the default `unsupportedToJpeg` converter will convert any unsupported format to jpeg.
Then the `jpegToPng` converter will convert any jpeg to png.

**Review by OmerFlame — APPROVED** [review]
Sorry for taking so long with this. This is a great improvement!

**Pull request merged**
